20240912-R-scheduler-auto-scale.md

```diff
@@ -0,0 +1,110 @@
+instance via a ClusterIP service, rather than needing to manually track and connect to each instance as in the current 
+headless service model. This `WatchHosts` RPC dynamically provides the current set of active Scheduler instances to the 
+sidecar, keeping it updated as Schedulers scale up or down.
+
+- For **backward compatibility**, we will check to determine whether the `WatchHosts` RPC is available or not. If it 
+exists, Dapr will use it to manage Scheduler connections dynamically. If not, Dapr will continue using the existing 
+static connection logic, maintaining compatibility with earlier versions by sending the full list back of available 
+Schedulers.
+
+```protobuf
+service Scheduler {
+  ...
+  rpc WatchHosts(WatchHostsRequest) returns (stream WatchHostsResponse) {}
+}
+
+message HostPort {
+  string host = 1;
+  uint32 port = 2;
+}
+
+message WatchHostsResponse {
+  repeated HostPort hosts = 1;
+}
+
+message WatchHostsRequest {}
+
+```
+
+3. **1/3/5 Schedulers with respective embedded etcd instances, beyond 5 instances run Schedulers stateless**. We want to
+optionally run embedded etcd in Scheduler for cases where it makes sense. Cases to consider:
+- running 1 Scheduler means running it with embedded etcd
+- running 2 Schedulers means running the first with embedded etcd, the second Scheduler will be stateless and connect to
+the existing etcd
+- running 3 Schedulers means running all with embedded etcd
+- running 4 Schedulers means running 3 Schedulers with embedded etcd, and the last one will be stateless and connect to
+an existing etcd. Note, here there is no gain to round-robin-ing between the existing etcd instances due to the way 
+etcd works with raft under the hood.
+- running 5 Schedulers means running 5 Schedulers with embedded etcd.
+- running beyond 5 Schedulers means running 5 Schedulers with embedded etcd, and any Scheduler beyond 5 will be 
+stateless and connect to an existing etcd instance.
+
+![Stateless Schedulers Example](./resources/20240912-R-scheduler-auto-scale/statelessSchedulersExample.png)
+
+etcd must run with an odd number of replicas. In dapr, 5 will be our practical limit while running inside Schedulers. 
+With the cron scheduler library and embedded etcd already separate processes in Scheduler, we just need to optionally 
+run the embedded etcd.
+
+Regarding [etcd's cluster reconfiguration operations guide](https://etcd.io/docs/v3.3/op-guide/runtime-configuration/#cluster-reconfiguration-operations),
+for adding members, or etcd instances, we will need to set the `initial-cluster-state` to be existing for the new 
+instances. We do not need to update the configuration of existing members manually, as once the new members join the 
+cluster, it will automatically be integrated into the consensus and the cluster's state will include the new member. 
+
+As noted in their documentation, we _**should**_ set the `strict-reconfig-check` flag on etcd such that we reject
```

Comment:
We still need to call `etcdctl member remove a8266ecf031671f3` to remove on scaling down. I suggest we get out of the business of auto-scaling etcd and run in 2 modes: 1 or 3.